gapic/src/main/com/google/gapid/perfetto/models/FrameInfo.java

```diff
@@ -31,70 +37,308 @@
+    //              (select distinct track_id from frame_slice where layer_name = 'com.xx.Layer1') t
+    //          on gpu_track.id = t.track_id);
+    //      where name GLOB 'APP_*')) phase
+    String displayQuery =
+        displayPhaseQuery(trackIdQuery(layerTracksQuery(layerName),"Display_*"));
+    String appQuery = phaseQuery(trackIdQuery(layerTracksQuery(layerName), "APP_*"));
+    String gpuQuery = phaseQuery(trackIdQuery(layerTracksQuery(layerName), "GPU_*"));
+    String compositionQuery = phaseQuery(trackIdQuery(layerTracksQuery(layerName), "SF_*"));
+    String displayViewName = baseName + "_DISPLAY";
+    String appViewName = baseName + "_APP";
+    String gpuViewName = baseName + "_GPU";
+    String compositionViewName = baseName + "_COMPOSITION";
 
-            // Sort buffers by name, the query is sorted by track id.
-            buffers.sort((b1, b2) -> {
-              // Displayed Frame track should always be at top
-              if (b1.name.equals(DISPLAYED_FRAME_TRACK_NAME)) {
-                return -1;
-              } else if(b2.name.equals(DISPLAYED_FRAME_TRACK_NAME)) {
-                return 1;
-              }
-              return b1.name.compareTo(b2.name);
+    // we need to determine the max depth of these views here, so that panel creation
+    // can happen appropriately.
+    return transformAsync(qe.queries(
+        dropView(appViewName),
+        dropView(gpuViewName),
+        dropView(compositionViewName),
+        dropView(displayViewName),
+        createView(appViewName, appQuery),
+        createView(gpuViewName, gpuQuery),
+        createView(compositionViewName, compositionQuery),
+        createView(displayViewName, displayQuery)), $ -> {
+          return transformAsync(getMaxDepth(qe, displayViewName), displayDepth -> {
+            currentLayerPhases.add(new Event("On Display",baseName + "_DISPLAY", displayDepth + 1,
+                DISPLAY_TOOLTIP));
+            return transformAsync(getMaxDepth(qe, appViewName), appDepth -> {
+              currentLayerPhases.add(new Event("Application", baseName + "_APP", appDepth + 1,
+                  APP_TOOLTIP));
+              return transformAsync(getMaxDepth(qe, gpuViewName), gpuDepth -> {
+                currentLayerPhases.add(new Event("Wait for GPU", baseName + "_GPU", gpuDepth + 1,
+                    GPU_TOOLTIP));
+                return transformAsync(getMaxDepth(qe, compositionViewName), compositionDepth -> {
+                  currentLayerPhases.add(new Event("Composition", baseName + "_COMPOSITION",
+                      compositionDepth + 1, COMPOSITION_TOOLTIP));
+                  phases.add(currentLayerPhases);
+                  if (idx + 1 >= layerNames.size()) {
+                    return Futures.immediateFuture(phases);
+                  }
+                  return createPhases(qe, phases, layerNames, idx + 1);
+                });
+              });
             });
```

Comment:
I'm a little worried about the 9 recursive transform function calling here, if they don't have time dependency, can we use some utility methods like `Futures.allAsList` to make it iterative rather than recursive? Not extremely sure, may need some suggestions from Pascal here.

Reply:
Good suggestion. I did not know about allAsList. I've updated the function now to use this. PTAL